packages/tpetra/core/src/Tpetra_Details_DistributorPlan.cpp

```diff
@@ -0,0 +1,975 @@
+    return "By createReverseDistributor";
+  case Details::DISTRIBUTOR_INITIALIZED_BY_COPY:
+    return "By copy constructor";
+  default:
+    return "INVALID";
+  }
+}
+
+DistributorPlan::DistributorPlan(Teuchos::RCP<const Teuchos::Comm<int>> comm)
+  : comm_(comm),
+    howInitialized_(DISTRIBUTOR_NOT_INITIALIZED),
+    reversePlan_(Teuchos::null),
+    sendType_(DISTRIBUTOR_SEND),
+    barrierBetweenRecvSend_(barrierBetween_default),
+    useDistinctTags_(useDistinctTags_default),
+    sendMessageToSelf_(false),
+    numSendsToOtherProcs_(0),
+    maxSendLength_(0),
+    numReceives_(0),
+    totalReceiveLength_(0)
+{ }
+
+DistributorPlan::DistributorPlan(const DistributorPlan& otherPlan)
+  : comm_(otherPlan.comm_),
+    howInitialized_(DISTRIBUTOR_INITIALIZED_BY_COPY),
+    reversePlan_(otherPlan.reversePlan_),
+    sendType_(otherPlan.sendType_),
+    barrierBetweenRecvSend_(otherPlan.barrierBetweenRecvSend_),
+    useDistinctTags_(otherPlan.useDistinctTags_),
+    sendMessageToSelf_(otherPlan.sendMessageToSelf_),
+    numSendsToOtherProcs_(otherPlan.numSendsToOtherProcs_),
+    procIdsToSendTo_(otherPlan.procIdsToSendTo_),
+    startsTo_(otherPlan.startsTo_),
+    lengthsTo_(otherPlan.lengthsTo_),
+    maxSendLength_(otherPlan.maxSendLength_),
+    indicesTo_(otherPlan.indicesTo_),
+    numReceives_(otherPlan.numReceives_),
+    totalReceiveLength_(otherPlan.totalReceiveLength_),
+    lengthsFrom_(otherPlan.lengthsFrom_),
+    procsFrom_(otherPlan.procsFrom_),
+    startsFrom_(otherPlan.startsFrom_),
+    indicesFrom_(otherPlan.indicesFrom_)
+{ }
+
+int DistributorPlan::getTag(const int pathTag) const {
+  return useDistinctTags_ ? pathTag : comm_->getTag();
+}
+
+size_t DistributorPlan::createFromSends(const Teuchos::ArrayView<const int>& exportProcIDs) {
+  using Teuchos::outArg;
+  using Teuchos::REDUCE_MAX;
+  using Teuchos::reduceAll;
+  using std::endl;
+
+  const size_t numExports = exportProcIDs.size();
+  const int myProcID = comm_->getRank();
+  const int numProcs = comm_->getSize();
+
+  // exportProcIDs tells us the communication pattern for this
+  // distributor.  It dictates the way that the export data will be
+  // interpreted in doPosts().  We want to perform at most one
+  // send per process in doPosts; this is for two reasons:
+  //   * minimize latency / overhead in the comm routines (nice)
+  //   * match the number of receives and sends between processes
+  //     (necessary)
+  //
+  // Teuchos::Comm requires that the data for a send are contiguous
+  // in a send buffer.  Therefore, if the data in the send buffer
+  // for doPosts() are not contiguous, they will need to be copied
+  // into a contiguous buffer.  The user has specified this
+  // noncontiguous pattern and we can't do anything about it.
+  // However, if they do not provide an efficient pattern, we will
+  // warn them if one of the following compile-time options has been
+  // set:
+  //   * HAVE_TPETRA_THROW_EFFICIENCY_WARNINGS
```

Comment:
Here's an untested feature.  I didn't know it existed.
Eventually (not this PR), we should either make the feature easier to enable (e.g., via environment variable) or get rid of it.

Reply:
agreed, will add this to my list of things to investigate.